library/std/src/sys/pal/windows/api.rs

```diff
@@ -34,6 +34,105 @@ use core::ptr::addr_of;
 
 use super::c;
 
+/// Creates a null-terminated UTF-16 string from a str.
+macro_rules! wide_str {
+    ($str:literal) => {{
+        const _: () = {
+            if core::slice::memchr::memchr(0, $str.as_bytes()).is_some() {
+                panic!("null terminated strings cannot contain interior nulls");
+            }
+        };
+        utf16!(concat!($str, '\0'))
+    }};
+}
+
+/// Creates a UTF-16 string from a str without null termination.
+macro_rules! utf16 {
+    // Note: this macro uses triple underscores to avoid const cycles
+    ($str:expr) => {{
+        const ___UTF8: &str = $str;
+        const ___UTF16_LEN: usize = crate::sys::pal::windows::api::utf16_len(___UTF8);
+        const ___UTF16: [u16; ___UTF16_LEN] = crate::sys::pal::windows::api::to_utf16(___UTF8);
+        &___UTF16
+    }};
```

Comment:
huh. const cycles...?

Reply:
I mean like:

```rust
const UTF8: &str = "hello";
{
    const UTF8: &str = UTF8;
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66187861d2fa095ad67c03359b6d097b)

Reply:
That would be sensible but... https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85373b84b46ca4d49dd9544030fdc081

Reply:
Oh, so it's only a problem if we add it as an argument.

Reply:
Yeah. I guess there's the option of saying "don't do that" or requiring it to be a literal. But the workaround is relatively painless as macros go.